File: src/webui/src/services/sources.ts
import { authFetch, getApiUrl } from "./api";
// Sources service for LDF server communication


export interface Source {
  id: string;
  name: string;
  url: string;
  component_ids: string[];
  retrieval_method: string;
  url_template?: string;
  forge_type: string;
  version_filter?: string;
  default_version?: string;
  priority: number;
  enabled: boolean;
  is_system: boolean;
  owner_id?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateSourceRequest {
  name: string;
  url: string;
  component_ids?: string[];
  retrieval_method?: string;
  url_template?: string;
  forge_type?: string;
  version_filter?: string;
  default_version?: string;
  priority?: number;
  enabled?: boolean;
  is_system?: boolean;
}

export interface UpdateSourceRequest {
  name?: string;
  url?: string;
  component_ids?: string[];
  retrieval_method?: string;
  url_template?: string;
  forge_type?: string;
  version_filter?: string;
  default_version?: string;
  priority?: number;
  enabled?: boolean;
}

export type ListResult =
  | { success: true; sources: Source[] }
  | {
      success: false;
      error:
        | "unauthorized"
        | "network_error"
        | "not_configured"
        | "internal_error";
      message: string;
    };

export type GetResult =
  | { success: true; source: Source }
  | {
      success: false;
      error:
        | "not_found"
        | "forbidden"
        | "unauthorized"
        | "network_error"
        | "not_configured"
        | "internal_error";
      message: string;
    };

export type CreateResult =
  | { success: true; source: Source }
  | {
      success: false;
      error:
        | "conflict"
        | "unauthorized"
        | "forbidden"
        | "invalid_request"
        | "network_error"
        | "not_configured"
        | "internal_error";
      message: string;
    };

export type UpdateResult =
  | { success: true; source: Source }
  | {
      success: false;
      error:
        | "not_found"
        | "forbidden"
        | "unauthorized"
        | "invalid_request"
        | "network_error"
        | "not_configured"
        | "internal_error";
      message: string;
    };

export type DeleteResult =
  | { success: true }
  | {
      success: false;
      error:
        | "not_found"
        | "forbidden"
        | "unauthorized"
        | "network_error"
        | "not_configured"
        | "internal_error";
      message: string;
    };

// List all sources (system + user sources) for the current user
export async function listSources(): Promise<ListResult> {
  const url = getApiUrl("/sources");

  if (!url) {
    return {
      success: false,
      error: "not_configured",
      message: "Server connection not configured",
    };
  }

  try {
    const result = await authFetch(url);

    if (result.ok) {
      const data = result.data as any;
      const sources: Source[] = data.sources || [];
      return { success: true, sources };
    }

    if (result.status === 401) {
      return {
        success: false,
        error: "unauthorized",
        message: "Authentication required",
      };
    }

    return {
      success: false,
      error: "internal_error",
      message: "Failed to fetch sources",
    };
  } catch (err) {
    return {
      success: false,
      error: "network_error",
      message:
        err instanceof Error ? err.message : "Failed to connect to server",
    };
  }
}

// Get a source by ID
export async function getSourceById(id: string): Promise<GetResult> {
  const url = getApiUrl(`/sources/${id}`);

  if (!url) {
    return {
      success: false,
      error: "not_configured",
      message: "Server connection not configured",
    };
  }

  try {
    const result = await authFetch(url);

    if (result.ok) {
      const source = result.data as any;
      return { success: true, source };
    }

    if (result.status === 401) {
      return {
        success: false,
        error: "unauthorized",
        message: "Authentication required",
      };
    }

    if (result.status === 403) {
      return {
        success: false,
        error: "forbidden",
        message: "You do not have access to this source",
      };
    }

    if (result.status === 404) {
      return {
        success: false,
        error: "not_found",
        message: "Source not found",
      };
    }

    return {
      success: false,
      error: "internal_error",
      message: "Failed to fetch source",
    };
  } catch (err) {
    return {
      success: false,
      error: "network_error",
      message:
        err instanceof Error ? err.message : "Failed to connect to server",
    };
  }
}

// Create a source (user source, or system source if is_system=true and user is admin)
export async function createSource(
  request: CreateSourceRequest,
): Promise<CreateResult> {
  const url = getApiUrl("/sources");

  if (!url) {
    return {
      success: false,
      error: "not_configured",
      message: "Server connection not configured",
    };
  }

  try {
    const result = await authFetch(url, {
      method: "POST",
      body: JSON.stringify(request),
    });

    if (result.ok) {
      const source = result.data as any;
      return { success: true, source };
    }

    if (result.status === 401) {
      return {
        success: false,
        error: "unauthorized",
        message: "Authentication required",
      };
    }

    if (result.status === 403) {
      return {
        success: false,
        error: "forbidden",
        message: "Admin access required for system sources",
      };
    }

    if (result.status === 400) {
      return {
        success: false,
        error: "invalid_request",
        message: "Invalid source data",
      };
    }

    if (result.status === 409) {
      return {
        success: false,
        error: "conflict",
        message: "A source with this URL already exists",
      };
    }

    return {
      success: false,
      error: "internal_error",
      message: "Failed to create source",
    };
  } catch (err) {
    return {
      success: false,
      error: "network_error",
      message:
        err instanceof Error ? err.message : "Failed to connect to server",
    };
  }
}

// Update a source by ID
export async function updateSource(
  id: string,
  request: UpdateSourceRequest,
): Promise<UpdateResult> {
  const url = getApiUrl(`/sources/${id}`);

  if (!url) {
    return {
      success: false,
      error: "not_configured",
      message: "Server connection not configured",
    };
  }

  try {
    const result = await authFetch(url, {
      method: "PUT",
      body: JSON.stringify(request),
    });

    if (result.ok) {
      const source = result.data as any;
      return { success: true, source };
    }

    if (result.status === 401) {
      return {
        success: false,
        error: "unauthorized",
        message: "Authentication required",
      };
    }

    if (result.status === 403) {
      return {
        success: false,
        error: "forbidden",
        message: "You do not have permission to update this source",
      };
    }

    if (result.status === 404) {
      return {
        success: false,
        error: "not_found",
        message: "Source not found",
      };
    }

    if (result.status === 400) {
      return {
        success: false,
        error: "invalid_request",
        message: "Invalid source data",
      };
    }

    return {
      success: false,
      error: "internal_error",
      message: "Failed to update source",
    };
  } catch (err) {
    return {
      success: false,
      error: "network_error",
      message:
        err instanceof Error ? err.message : "Failed to connect to server",
    };
  }
}

// Delete a source by ID
export async function deleteSource(id: string): Promise<DeleteResult> {
  const url = getApiUrl(`/sources/${id}`);

  if (!url) {
    return {
      success: false,
      error: "not_configured",
      message: "Server connection not configured",
    };
  }

  try {
    const result = await authFetch(url, { method: "DELETE" });

    if (result.ok) {
      return { success: true };
    }

    if (result.status === 401) {
      return {
        success: false,
        error: "unauthorized",
        message: "Authentication required",
      };
    }

    if (result.status === 403) {
      return {
        success: false,
        error: "forbidden",
        message: "You do not have permission to delete this source",
      };
    }

    if (result.status === 404) {
      return {
        success: false,
        error: "not_found",
        message: "Source not found",
      };
    }

    return {
      success: false,
      error: "internal_error",
      message: "Failed to delete source",
    };
  } catch (err) {
    return {
      success: false,
      error: "network_error",
      message:
        err instanceof Error ? err.message : "Failed to connect to server",
    };
  }
}

File: src/ldfctl/internal/client/sources.go
package client

import (
	"context"
	"fmt"
)

// Source represents a source resource
type Source struct {
	ID             string `json:"id"`
	Name           string `json:"name"`
	URL            string `json:"url"`
	ForgeType      string `json:"forge_type"`
	ComponentID    string `json:"component_id"`
	IsSystem       bool   `json:"is_system"`
	OwnerID        string `json:"owner_id"`
	VersionFilter  string `json:"version_filter"`
	LastSyncAt     string `json:"last_sync_at"`
	LastSyncStatus string `json:"last_sync_status"`
	VersionCount   int    `json:"version_count"`
	CreatedAt      string `json:"created_at"`
	UpdatedAt      string `json:"updated_at"`
}

// SourceListResponse represents a list of sources
type SourceListResponse struct {
	Count   int      `json:"count"`
	Sources []Source `json:"sources"`
}

// SourceVersion represents a discovered version
type SourceVersion struct {
	Version string `json:"version"`
	URL     string `json:"url"`
	Type    string `json:"type"`
}

// SourceVersionListResponse represents a list of versions
type SourceVersionListResponse struct {
	Count    int             `json:"count"`
	Versions []SourceVersion `json:"versions"`
}

// SyncTriggerResponse represents the response from triggering a sync
type SyncTriggerResponse struct {
	Message  string `json:"message"`
	SourceID string `json:"source_id"`
	Status   string `json:"status"`
}

// SyncStatusResponse represents the sync status
type SyncStatusResponse struct {
	SourceID     string `json:"source_id"`
	Status       string `json:"status"`
	LastSyncAt   string `json:"last_sync_at"`
	VersionCount int    `json:"version_count"`
	Error        string `json:"error,omitempty"`
}

// CreateSourceRequest represents the request to create a source
type CreateSourceRequest struct {
	Name          string `json:"name"`
	URL           string `json:"url"`
	ComponentID   string `json:"component_id"`
	VersionFilter string `json:"version_filter,omitempty"`
}

// UpdateSourceRequest represents the request to update a source
type UpdateSourceRequest struct {
	Name          string `json:"name,omitempty"`
	URL           string `json:"url,omitempty"`
	VersionFilter string `json:"version_filter,omitempty"`
}

// ListSources returns all sources
func (c *Client) ListSources(ctx context.Context) (*SourceListResponse, error) {
	var resp SourceListResponse
	if err := c.Get(ctx, "/v1/sources", &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetSource returns a single source by ID
func (c *Client) GetSource(ctx context.Context, id string) (*Source, error) {
	var resp Source
	if err := c.Get(ctx, fmt.Sprintf("/v1/sources/%s", id), &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// CreateSource creates a new user source
func (c *Client) CreateSource(ctx context.Context, req *CreateSourceRequest) (*Source, error) {
	var resp Source
	if err := c.Post(ctx, "/v1/sources", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// UpdateSource updates an existing source
func (c *Client) UpdateSource(ctx context.Context, id string, req *UpdateSourceRequest) (*Source, error) {
	var resp Source
	if err := c.Put(ctx, fmt.Sprintf("/v1/sources/%s", id), req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// DeleteSource deletes a source
func (c *Client) DeleteSource(ctx context.Context, id string) error {
	return c.Delete(ctx, fmt.Sprintf("/v1/sources/%s", id), nil)
}

// ListSourceVersions returns discovered versions for a source
func (c *Client) ListSourceVersions(ctx context.Context, id string) (*SourceVersionListResponse, error) {
	var resp SourceVersionListResponse
	if err := c.Get(ctx, fmt.Sprintf("/v1/sources/%s/versions", id), &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// SyncSource triggers a version sync for a source
func (c *Client) SyncSource(ctx context.Context, id string) (*SyncTriggerResponse, error) {
	var resp SyncTriggerResponse
	if err := c.Post(ctx, fmt.Sprintf("/v1/sources/%s/sync", id), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetSyncStatus returns the sync status for a source
func (c *Client) GetSyncStatus(ctx context.Context, id string) (*SyncStatusResponse, error) {
	var resp SyncStatusResponse
	if err := c.Get(ctx, fmt.Sprintf("/v1/sources/%s/sync/status", id), &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ClearSourceVersions deletes all cached versions for a source
func (c *Client) ClearSourceVersions(ctx context.Context, id string) error {
	return c.Delete(ctx, fmt.Sprintf("/v1/sources/%s/versions", id), nil)
}
